metasrv/src/meta_service/meta_leader.rs

```diff
@@ -151,6 +156,40 @@ impl<'a> MetaLeader<'a> {
         self.change_membership(membership).await
     }
 
+    /// A node leave the cluster.
+    ///
+    /// - Remove the node from cluster.
+    /// - Remove the node from membership.
+    ///
+    /// If the node is not in cluster membership, it still returns Ok.
+    #[tracing::instrument(level = "debug", skip(self))]
+    pub async fn leave(&self, req: LeaveRequest) -> Result<(), MetaError> {
+        let node_id = req.node_id;
+        let metrics = self.meta_node.raft.metrics().borrow().clone();
+        let membership = metrics.membership_config.membership.clone();
+
+        if !membership.contains(&node_id) {
+            return Ok(());
+        }
+
+        // TODO(xp): deal with joint config
+        assert!(membership.get_ith_config(1).is_none());
+
+        // safe unwrap: if the first config is None, panic is the expected behavior here.
+        let mut membership = membership.get_ith_config(0).unwrap().clone();
+
+        membership.remove(&node_id);
+
+        let ent = LogEntry {
+            txid: None,
+            cmd: Cmd::RemoveNode { node_id },
+        };
+
+        self.write(ent.clone()).await?;
+
+        self.change_membership(membership).await
```

Comment:
If the first entry that removes a node is committed and applied but the second membership entry is not, the the network will find a absent node, which will cause a network error.